### DISCUSSION TITLE
background location modal with createBrowserRouter

### DISCUSSION THREAD

**Discussion by ZachMayry** [remix-run/react-router#9864] [Proposals]
### What is the new or updated feature that you are suggesting?

Somehow add support for a background location with routes
At first glance I'm not understanding how this modal example would work with the new data routers.
https://github.com/remix-run/react-router/blob/dev/examples/modal/src/App.tsx#L56-L70
If this feature is supported, it would be nice to have the modal example updated

### Why should this feature be included?

Allowing modal routes to be on top of background location routes is a feature our application is built on but we would like to use the new data loading features as well. 
It would be nice to not have to choose data loading or background locations for modals.

**Comment by alexgorbatchev** [comment]
I personally feel that the way the modal pattern is handled in react-router is basically a very inconvenient patch-job. It's a very common pattern for modals/drawers/etc to be visible at the same time as originating content. It was possible to do this with v5 relatively easy. Migrating to v6 isn't easy to begin with. The modal pattern deserves proper attention.

```tsx
        // react-router@5
        <Route path={path}>
          <Content />
          <Route path={`${path}/new`}><Overlay /></Route>
        </Route>
```

**Comment by alexgorbatchev** [comment]
This took some times to figure out, hope it will help someone. To get basic modal functionality in v6 you can do something like this:

```tsx
<Routes>
  <Route
    path=":id/grants/*"
    element={
      <>
        <Content />
        <Routes>
          <Route path="new" element={<Overlay />} />
        </Routes>
      </>
    }
  />
</Routes>
```

The [modal example](https://github.com/remix-run/react-router/blob/main/examples/modal/src/App.tsx) assumes you want to display modal content full screen when navigating directly to the route. IMO this isn't a typical modal pattern. If I open a side drawer and want to share the link, I don't want that drawer to be full screen when my friend navigates to my link.

**Comment by ZachMayry** [comment]
Modals don't have to be fullscreen. I started with the example in an application I'm working on it's working as you'd expect (not full screen). The CSS sets the dimensions and any open space displays the background route.

**Comment by alexgorbatchev** [comment]
@ZachMayry by "full screen" i didn't mean visually full screen. In the [example](https://github.com/remix-run/react-router/blob/main/examples/modal/src/App.tsx) if you navigate directly to `/img/:id` you see the image only, but if you click on a thumbnail you are taken to `/img/:id` but you see a modal.

**Comment by brophdawg11** [comment]
I'm going to convert this to a discussion so it can go through our new [Open Development](https://remix.run/blog/open-development) process.  Please upvote the new Proposal if you'd like to see this considered!

**Comment by xavier-lc** [comment]
I'm currently using two different sets of routes on my app. They are rendered under a `HistoryRouter`, and I'm taking advantage of the second argument on `useRoutes` to scope their location. Something like:

```
const baseLocation = useRef(...);
const modalLocation = useRef(...);

<HistoryRouter history={history}>
  {useRoutes(baseRoutes, baseLocation)}
  {useRoutes(modalRoutes, modalLocation)}
</HistoryRouter>
```

Now I'm looking to replace that structure with a data-router using `createBrowserRouter`. Is there a way to keep the scoped location? Or I should keep that in a context and check inside each component?

**Comment by AnthonyCrowcroft** [comment]
surely for data routes to be fully functional this needs to be possible?

**Comment by doubleChu** [comment]
I found this example using createBrowserRouter: https://stackblitz.com/edit/github-9lelz2-tdcovz?file=src%2Fmain.tsx.
It has a problem that when navigating to "/img/:id", background page would change as the modal dialog appears.
```tsx
<RouterProvider router={router} />
```
 acts like
 ```tsx
<BrowserRouter>
    <Routes>...</Routes>
</BrowserRouter>
```
in the official example,
```tsx
{state?.backgroundLocation && (
    <Routes>
       <Route path="/img/:id" element={<Modal />} />
    </Routes>
)}
```
is inserted into before \</BrowserRouter>.
But in the stackblitz link above, it works like it's inserted after it.
So if I am not in a function component when using ```createBrowserRouter```, how can I achieve this?

**Comment by bvanderplanck** [comment]
I have stumbled into the same issue, using `createBrowserRouter` to take advantage of the loaders. We have a dynamic app where the route can either be a modal OR a regular page depending on a setting.
I ended up doing this:

```
import { UNSAFE_RouteContext, useLocation } from 'react-router-dom'
import { motion } from 'framer-motion'
import { useContext, useEffect, useState } from 'react'

const AppContent = () => {
  const location = useLocation()
  const context = useContext(UNSAFE_RouteContext)
  const [ innerLocation, setInnerLocation ] = useState<string>('')
  const [ outlet, setOutlet ] = useState<React.ReactElement | null>(context.outlet)

  useEffect(() => {
    if (location.state?.modal) {
      return
    }

    setInnerLocation(location.key)
    setOutlet(context.outlet)
  }, [ location ])

  return (
    <motion.div
      key={innerLocation}
      initial={{ y: 10, opacity: 0 }}
      animate={{ y: 0, opacity: 1 }}
      exit={{ y: -10, opacity: 0 }}
      transition={{ duration: 0.2 }}
    >
      {outlet}
    </motion.div>
  )
}

export default AppContent
```

Passing the location state on the Link element conditionally.
The modal itself is injected in the DOM through portals and context. But at least I managed changing the URL in the history without changing the actual page state which is what I was aiming for. If that can help anyone..

**Reply by gvozdenkov** [reply]
@vitorcamachoo @bvanderplanck @doubleChu @AnthonyCrowcroft @xavier-lc 
I had a task to make it so that from the root route, clicking on the product card would open a modal window with details about this product. But if the user directly entered the route address with details about the product, then a regular page with a description of the product would open.

I wanted to organize this using `createBrowserRouter` and the new react-router-dome trick with loader and action.

The difficulty is that we cannot use hooks at the createBrowserRouter creation level, so the previous background location techniques do not work.

I created such a routing scheme. The `ingredient/:id` route is just a modal window (and a regular page by condition, more on that below)
```js
export const router = createBrowserRouter([
  {
    path: PATH.HOME,
    element: <RootLayout />,
    loader: ingredientsLoader(queryClient),
    children: [
      {
        children: [
          {
            path: '/',
            element: <Home />,
            children: [
              {
                path: 'ingredients/:id',
                element: <IngredientModal />,
              },
            ],
          },

          {
            path: '/profile',
            element: <OnlyAuth component={<ProfileLayout />} />,
            action: ProfileLayout.logout(store.dispatch),
            loader: userLoader(queryClient),
              },
            ],
          },

          {
            path: '/login',
            element: <OnlyUnAuth component={<Login />} />,
            action: Login.login(store.dispatch),
          },
          {
            path: '/register',
            element: <OnlyUnAuth component={<Register />} />,
          },
        ],
      },
    ],
  },
]);
```

The Home component renders product cards. The links on these cards look like this:

```js
<Link
        to={`ingredients/${ingredient._id}`}
        state={{ from: location }}>
              {ingredient.name}
</Link>
```
I write down in the link information about where the transition is made on this link. If from the page Home - state will contain an object with information. If the modal route is navigated directly from the browser line, then the state object will be empty.

the Home component contains an Outlet component so that it can render a nested route with a modal in it:

```js
export const Home = () => {
  const { data: ingredients } = useQuery(ingredientsQuery());
  const location = useLocation();

  return (
    <section className="p-4">
      <ul>
        {ingredients.map((ingredient) => (
          <li key={ingredient._id}>
            <Link to={`ingredients/${ingredient._id}`} state={{ from: location }}>
              {ingredient.name}
            </Link>
          </li>
        ))}
      </ul>
      <Outlet />
    </section>
  );
};
```

It remains to make sure that the `IngredientModal` component understands what to render - a modal or a regular page. This is where it really gets tricky. In fact, `IngredientModal` always renders a modal. Just one modal looks like a regular modal - a window on top of the page with a translucent background. And the second modal fills the entire bottom page with a solid opaque color and the content is already rendered on top of this fill

```js
export const IngredientModal = () => {
  // get ingredients from TanStack query chache
  const { data: ingredients } = useQuery(ingredientsQuery());

 // get id from url params
  const { id } = useParams();
  const location = useLocation();
  
  // Check location.state content - if it's empty `isFromHome = false`
  const isFromHome = !!location.state;

  const ingredient = ingredients.find((ingredient) => ingredient._id === id);

  return isFromHome ? (
    <Modal>
      <div className="flex flex-col items-center max-w-lg mx-auto mt-10">
        <h1>{ingredient.name}</h1>
        <img src={ingredient.image_large} alt={ingredient.name} />
        <p>Price: ${ingredient.price}</p>
      </div>
    </Modal>
  ) : (
    <ModalFullScreen>
      <div className="flex flex-col items-center max-w-lg mx-auto mt-10">
        <h1>{ingredient.name}</h1>
        <img src={ingredient.image_large} alt={ingredient.name} />
        <p>Price: ${ingredient.price}</p>
      </div>
    </ModalFullScreen>
  );
};

```
`ModalFullScreen` component is a regular Modal, but the background is not transparent and there is no wrapper for the content that sets the dimensions of the modal and the alignment relative to the background

```js
export const ModalFullScreen = ({ children }) => {
  return createPortal(
    <Overlay>
      <Content>{children}</Content>
    </Overlay>,
    MODAL_PORTAL_EL
  );
};
```
And regular modal:

```js
export const Modal = ({ children }) => {
  const navigate = useNavigate();

  const handleClose = () => {
    navigate(-1);
  };

  return createPortal(
    <Overlay>
      <ModalWrapper>
          <button type="button" onClick={handleClose}>
            close
          </button>
        <Content>{children}</Content>
      </ModalWrapper>
    </Overlay>,
    document.querySelector('#modal-portal')
  );
};

```
Here is the result (there is no design, I test only functionality)
If click link on the home page '/'
![image](https://github.com/remix-run/react-router/assets/81020763/557956c3-a0ab-49dc-92e2-0df060cb0e06)

And if I directly past this adress in browser address bar I have this:
![image](https://github.com/remix-run/react-router/assets/81020763/3357e938-f6ff-4208-9e6a-4ee80e4e31e6)

**Reply by tanmayaBiswalOdiware** [reply]
@gvozdenkov So, if I understood correctly, the `<ul>` in the `Home` component is still getting rendered in the background but stays under the modal even when you put the address directly in the browser?

**Reply by gvozdenkov** [reply]
Heh.. Yep... But I don't know how to solve this task more clearly with router v6 data router:(

**Comment by vitorcamachoo** [comment]
Is there any clean way to have modal routes with data route API?

**Comment by Ashish-Krishna-K** [comment]
Anyone figured out a way to make this work?

**Comment by carrotpai** [comment]
after much wandering through the documentation, i came to a solution using matchRouters and renderMatches utils functions, but it looks more like bruteforce than a good solution.
```JSX
export const renderRoute = (
	routes: RouteObject[],
	location: string | Partial<Location>
): React.ReactElement | null => {
	const route = matchRoutes(routes, location);
	const backgroundPage = renderMatches(route);
	return backgroundPage;
};
```

i keep two router objects (one for modals and one for other routes).

```JSX
export const routes = createRoutesFromElements(
	<Route path="/" element={<RootLayout />}>
		<Route element={<PagesLayout />}>
			<Route index element={<MainPage />} />
			<Route path="blogs" element={<BlogsPage />} />
			<Route path="blog/:id" element={<BlogPage />} />
			<Route path="createPost" element={<CreatePostPage />} />
			<Route path="user/:id" element={<UserPage />} />
		</Route>
		<Route element={<AuthLayout />}>
			<Route path="login" element={<Login />} />
			<Route path="register" element={<Register />} />
		</Route>
	</Route>
);

export const modalsRoutes = createRoutesFromElements(
	<Route path="/">
		<Route
			path="user/:id"
			element={
				<Modal isModalOpened>
					<p>Modal With User</p>
				</Modal>
			}
		/>
		<Route
			path="login"
			element={
				<Modal isModalOpened>
					<p>Must be logged in</p>
				</Modal>
			}
		/>
	</Route>
);
```

then in root layout component

```JSX
function RootLayout() {
	const location = useLocation();

	if (location.state?.background) {
		const backgroundPage = renderRoute(
			routes,
			location.state.background
		);
		const modal = renderRoute(modalsRoutes, location.pathname);
		return (
			<div className={styles.content}>
				{backgroundPage}
				{modal} {/** portal */}
			</div>
		);
	}

	return (
		<div className={styles.content}>
			<Outlet />
		</div>
	);
}
```

I haven't done any data router features testing, this solution was made only with the goal of recreating the old functionality of modal routes

**Reply by gvozdenkov** [reply]
Hmmmm! Cool idea! Will play with it!

**Reply by nitzanashi** [reply]
@carrotpai How did you register the routes with `createBrowserRouter` in your example you created 2 routes objects but do you register them both like
```js
return <RouterProvider router={createBrowserRouter([...routes, ... modalsRoutes])} />;
```

Or is it something else?

A problem I encountered is that the page behind is being completely re-rendered so any local state would be lost.

**Reply by lxcid** [reply]
this looks the most elegant, i’m wondering when you branch off to do your own renderRoute, does it double render the RootLayout?

**Reply by ali-idrizi** [reply]
Using `renderMatches()` with data routes beats the purpose of using data routes in the first place. That's because data APIs do not work at all (e.g. if you had `errorElement` defined in one of the `modalsRoutes`, and a modal threw an error, it would not fallback to render `errorElement`). At this point you're better off just using `<BrowserRouter>` and not relying on the data APIs.

**Comment by productmap** [comment]
Two ways to make it https://devsday.ru/blog/details/114436

**Reply by gvozdenkov** [reply]
Yep, but this is not working if you need this double behavior for modals on / (home) route :(

**Comment by afkcodes** [comment]
have been struggling with the same i want to overlay the modal over the content which is rendered using a context but, the moment modal comes up on a route the outlet data goes off any suggestions or patterns how to fix this i am using routerProvider with createbrowserRouter

**Reply by gvozdenkov** [reply]
Can you show some sceens or video?

**Reply by afkcodes** [reply]
wanted to know if this is behaviour possible with RouterProvider i saw some examples with <Routes>

**Reply by gvozdenkov** [reply]
Sorry, I don't understand what behaviour do you mean...

**Reply by afkcodes** [reply]
@gvozdenkov hey sorry, completely missed this. I am trying to build something like spotify where when you click on the miniplayer below it comes as a overlay, what i am trying to do is bring that overlay as a route using react-router so that when i hit back it goes to the previous screen. also at the same time when you swipe down the player you can see the page content that was there. I dont know if i can post a screen record of spotify as a reference here.

**Reply by gvozdenkov** [reply]
@afkcodes Now I understand. You can do this. But when you swipe (close popup ovelay) it will mean going to the previous route. This can be done with close handler in popup overlay:

```js
 const handleClose = () => {
    navigate(-1);
  };
```

I realized this in [my study project](https://github.com/gvozdenkov/space-food). [IngredientModal component](https://github.com/gvozdenkov/space-food/blob/main/src/pages/ingredient-modal/ingredient-modal.tsx)

**Reply by afkcodes** [reply]
will look into this thanks

**Comment by OliverJAsh** [comment]
At Unsplash this issue is preventing us from migrating to the new data router.

@brophdawg11 This discussion has a lot of upvotes now. What's your stance on this issue? Is this something React Router plans to support again?

**Comment by a463** [comment]
Would also appreciate an update on this, it's an old issue and the 7th most upvoted discussion, so clearly there's interest. It sucks that we have to choose between data router or modal routes.

**Comment by haitaojarvis** [comment]
More than a year has passed, is there any progress in this discussion so far? We'd really like to see an official solution built in, something as simple and easy as Intercepting Routes in Next.js or Route Masking in TanStack Router. Thanks a lot!

**Comment by Dawar-codes** [comment]
No solution so far. I tried to render the Modal conditionally without giving it a route which works. But when I am on a different page and open the Modal the background location changes to where the Modal is rendering. Don't know where to render the modal so that it doesn't force the background location to change.

**Reply by joaolfern** [reply]
You should place your modal on the general layout, preferably with a portal, so it will be unaffected by navigating routes.

I'm using query parameters prefixed with a "modal-" keyword to condition the content. It fells This approach has required me to reimplement some basic routing functionality (ex: conditionally render component based current query parameter), but it was able to get the job done.


--
Edit: I've since realized my example about clearing query parameters on history back was incorrect. Browser history navigation works automatically when using query parameters for routing modals. Most of the work is on conditionally rendering each modal.

**Reply by nitzanashi** [reply]
I also following a similar approach to how @joaolfern described, having the Modal as a component on the general layout that conditionally renders itself when a Query Parameter is present

**Reply by Dawar-codes** [reply]
Well for that I have to make a new route for that modal component. Right?

**Reply by nitzanashi** [reply]
Not a new route, but you need to register the modal in some "MainLayout" Component
Something like:

```jsx
import {Modal} from './your-modal';


const MainLayout = () => {
    return (
        <Body>
            <Outlet />
            <Modal />
       </ Body>
    );
};


...

<Route element={<MainLayout/>}>
    <Route path="/home" element={<Home/>} />
</Route>
```

**Reply by Dawar-codes** [reply]
```
const router = createBrowserRouter([
  {
    path: "/",
    element: <RootLayout />,
    children: [
      {
        index: true,
        element: <Navigate to="/products" />,
      },
      {
        path: "products",
        element: <Products />,
      },
      {
        path: "products/:id",
        element: <ProductDetailPage />,
      },
      {
        path: "checkout",
        element: <Checkout />,
      },
    ],
  },
  {
    path: "/admin",
    element: <AdminLayout />, // Admin Layout for admin-specific pages
    children: [
      { path: "new", element: <NewProductPage /> }, // Route for adding new products
    ],
  },
  {
    path: "/auth",
    element: <AuthLayout />,
    children: [
      {
        path: "login",
        element: <LoginPage />,
      },
      {
        path: "signup",
        element: <SignupPage />,
      },
    ],
  },
]);

const queryClient = new QueryClient();

export default function App() {
  const modal = useSelector((state) => state.modal.open);
  useEffect(() => {
    console.log(modal);
    
  },[modal])
  return (
    <>
      
      <QueryClientProvider client={queryClient}>
        <RouterProvider router={router} >{modal && <CartPage />}</RouterProvider>
      </QueryClientProvider>
    </>
  );
}
```

this is my App.jsx where I define all the routes. As shown if I try to render it inside the Route Provider it doesn't work.

**Reply by joaolfern** [reply]
The modal should be placed on `RootLayout`

**Reply by Dawar-codes** [reply]
I tried but still get the same issue. My Root layout has an Outlet enclosed in header and footer. I tried it outside header and footer and also inside it. But neither works.

**Reply by Dawar-codes** [reply]
oh I sorted it out. was using navigate on onClose on the modal which was forcing it to navigate to main page.

**Reply by joaolfern** [reply]
I think it's because RooLayout is not really serving as a root.

Do something like this:

```ts
const router = createBrowserRouter([
  {
    path: "/",
    element: <RealRootLayout />,
    children: [
      {
        path: "/",
        element: <RootLayout />,
        children: [
          {
            index: true,
            element: <Navigate to="/products" />,
          },
          {
            path: "products",
            element: <Products />,
          },
          {
            path: "products/:id",
            element: <ProductDetailPage />,
          },
          {
            path: "checkout",
            element: <Checkout />,
          },
        ],
      },
      {
        path: "/admin",
        element: <AdminLayout />,
        children: [{ path: "new", element: <NewProductPage /> }],
      },
      {
        path: "/auth",
        element: <AuthLayout />,
        children: [
          {
            path: "login",
            element: <LoginPage />,
          },
          {
            path: "signup",
            element: <SignupPage />,
          },
        ],
      },
    ],
  },
]);

```

**Reply by Dawar-codes** [reply]
Hey thanks for the help Joaofern. The root layout was ok. As I said problem is now solved. Now the problem is just setting up onClose to behave correctly.

**Comment by OliverJAsh** [comment]
For reference, [there's a discussion on Discord here](https://discord.com/channels/770287896669978684/1362868896469881249) with @ryanflorence where he mentions plans to add a `rewrite` prop to links (similar to route masking in TanStack Router). That was 6 months ago now so I'm not sure when this is coming. 🤞

**Reply by brophdawg11** [reply]
I just added the `review` label so we can look over this at the next steering committee meeting.  Would you mind dropping some of the details from your convo with Ryan in here for reference?

**Reply by OliverJAsh** [reply]
Yay! ☺️

To quote Ryan:

> I'm thinking we add a `rewrite` prop to links for this
> 
> ```jsx
> <Link
>   to="/photo/123"
>   rewrite="?photo=123"
> />
> ```
> 
> And then your root layout does the same kind modal logic as it does today with the search params instead of the location state 
The X Platform (formerly twitter) does this too, where at literally any URL can go to /compose/post and it keeps the background there
Instead of you having to keep track of a "background location" and stuff, React Router would do it. It would use the rewrite as the location but give the browser the to location to display

**Reply by OliverJAsh** [reply]
I believe Ryan's idea is similar to [route masking in TanStack Router](https://tanstack.com/router/v1/docs/framework/react/guide/route-masking).

It might also be worth looking at how Next.js solves this problem. They take a different approach with ["parallel routes" and "intercepting routes"](https://nextjs.org/docs/app/api-reference/file-conventions/parallel-routes#modals).

**Reply by OliverJAsh** [reply]
In case it helps, here's a quick demo of how we use currently use modals + background locations at Unsplash (stuck on declarative mode as per above):

https://github.com/user-attachments/assets/7bc6eff7-6a3d-421e-8933-65d76135bb21

**Reply by brophdawg11** [reply]
We reviewed this at the steering committee meeting today and sounds like theres no immediate concerns with the `<Link rewrite>` API mentioned so we should be good to go.  Consider this a 👍 from me - just need to grab one more upvote from another SC member and we can open this up for community PRs in the interim, otherwise one of us will try to pick it up before too long.

**Comment by brophdawg11** [comment]
👋 Hey folks!  This is being tracked on the roadmap now via https://github.com/remix-run/react-router/issues/14717 and we have an alpha implementation in https://github.com/remix-run/react-router/pull/14716.

If anyone wants to do some alpha testing, there's an experimental release in `0.0.0-experimental-3fe27584d` that you can try out - please let us know what you think!

**Reply by OliverJAsh** [reply]
Great to see progress on this.

After playing with this for a little while, I'm starting to think that we might also need "parallel routes" [like Next.js](https://nextjs.org/docs/app/api-reference/file-conventions/parallel-routes#modals) and TanStack Router.

Imagine you have a route component that may be shown either inside a modal (soft navigation) or outside a modal (hard navigation), like Unsplash. In this case, you want the route's loader to run _including when the route appears inside the modal_.

**Reply by brophdawg11** [reply]
I think that might be out of scope for this RFC - the current thought for rewrite is just that you would [share utils](https://discord.com/channels/770287896669978684/1362868896469881249/1364640874935353456) to be able to run logic in a loader from the image route or when an image is in a modal ([PR example](https://github.com/remix-run/react-router/pull/14716/changes#diff-747555524b232c1b5fa58290cdf878a0aa8e577951d6590de491b4f36c4d511dR7)).

While this is a bit more manual - is there a technical limitation that doesn't allow you to build the UI you are looking for using rewriting?

**Reply by brophdawg11** [reply]
Related to [this comment](https://github.com/remix-run/react-router/pull/14716#issuecomment-3729654779), is the expectation that links are rewrite-aware when you're in a rewrite scenario?  I don't know if I quite follow how that would work - wouldn't you need to tell the router how to update _both_ the browser location and the rewritten router location?

For example:

* User in on `/gallery`
* Clicks `<Link to="/photo/1" rewrite="/gallery?photo=1">`
* The app opens Photo 1 in a modal overtop of `/gallery` (using the search param)
* If you wanted to render an edit link, wouldn't it need to:
  * Change the browser location to `/photo/1/edit`
  * Change the router location to something like `/gallery?photo=1&edit`

So it seems like the app would need to be explicit with something like `<Link to="/photo/1/edit" rewrite="/gallery?photo=1&edit">`?

**Reply by OliverJAsh** [reply]
> If you wanted to render an edit link, wouldn't it need to:
> * Change the browser location to `/photo/1/edit`
> * Change the router location to something like `/gallery?photo=1&edit`

In the example I gave in my original comment, the relative link adds another search param rather than changing the pathname, so it would be more like this:

* Change the browser location to `/photo/1?edit`
* Change the router location to something like `/gallery?photo=1&edit`

I think it's interesting to consider how this would work in Next.js and TanStack Router.

With Next.js I think this page/modal pattern would be implemented via "parallel routes" + "intercepting routes" as described [here](https://nextjs.org/docs/app/api-reference/file-conventions/parallel-routes#modals). The relative "edit" link just works because there's only one location to think about, i.e. the link would simply target `/photo/1?edit`.

As for TanStack Router, I believe this pattern would be implemented via "parallel routes" + "[_declarative_ route masking](https://tanstack.com/router/v1/docs/framework/react/guide/route-masking#declarative-route-masking)". The relative "edit" link just works because the route masking is applied automatically to all links, i.e. the link would simply target `/photo/1?edit`.

I could be wrong, I've never used either of these. I'm just going by the documentation.

**Reply by brophdawg11** [reply]
In the search param example - you would expect that the param is applied to both the browser URL and the router URL?  It looks like that is the case on unsplash where the `modal` param is added to both locations.  So it introduces a bit of tight coupling to links to be opened from within rewritten UIs.  Maybe that should the default behavior but a nested `Link` with both a `to`/`rewrite` would allow you to provide different URL updated within a rewrite UI.  I can play around with that pattern a bit more to see if I can better replicate what you are doing on unsplash.

**Comment by OliverJAsh** [comment]
Link masking has been released in 7.13.1: https://github.com/remix-run/react-router/blob/main/CHANGELOG.md#url-masking-unstable

I can think we can close this now?

**Reply by brophdawg11** [reply]
We usually keep these open until it's released as stable - it serves as a good place to continue discussion from early testers.  I added the Roadmap label and moved it along on the [board](https://github.com/orgs/remix-run/projects/5/views/1)

**Comment by brophdawg11** [comment]
Hey folks!  As noted above, we released `<Link unstable_mask>` in 7.13.1 yesterday which enables this UI pattern in data mode.  You can see the original example converted to the new pattern in Data Mode in [this example](https://github.com/remix-run/react-router/tree/main/examples/modal-data-router).  Please let us know if you give it a shot and have any questions/feedback!

**Comment by tarungunampalli1** [comment]
I'd love to see [declarative route masking](https://tanstack.com/router/v1/docs/guide/route-masking#declarative-route-masking) like in tanstack router. The ergonomics of specifying logic related to a specific route at _every_ navigation to that route isn't very good

**Reply by brophdawg11** [reply]
I would recommend a new RFC for this concept as this RFC was focused on the same behavior from declarative mode which required call-site background location management.  For the time being I think you could probably wire up a declarative version in a small wrapper `Link` component that inspected `useLocation` and `to`.

**Comment by brophdawg11** [comment]
Stabilized in 7.15.0
